calibrate-touchscreen/src/main.rs

```diff
@@ -0,0 +1,429 @@
+                        registry.bind::<wl_seat::WlSeat, _, _>(name, 1, qh, ());
+                    }
+                    "xdg_wm_base" => {
+                        let wm_base = registry.bind::<xdg_wm_base::XdgWmBase, _, _>(name, 1, qh, ());
+                        state.wm_base = Some(wm_base);
+                    }
+                    _ => {}
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+impl Dispatch<wl_output::WlOutput, ()> for State {
+    fn event(
+        state: &mut Self,
+        output: &wl_output::WlOutput,
+        event: wl_output::Event,
+        _: &(),
+        _: &Connection,
+        qh: &QueueHandle<Self>,
+    ) {
+        use wayland_client::protocol::wl_output::Mode;
+
+        match event {
+            wl_output::Event::Mode { flags: WEnum::Value(flags), width, height, refresh } => {
+                if flags.contains(Mode::Current) {
+                    state.outputs.entry(output.clone())
+                        .and_modify(|fullscreen| *fullscreen = Some(FullscreenSurface{
+                            width: width,
+                            height: height,
+                            wl_surface: None,
+                            buffer: None,
+                            xdg_surface: None,
+                    }));
+
+                    state.draw_fullscreen_surfaces(qh);
+                }
+            }
+
+            _ => {}
+        }
+    }
+}
+
+fn draw_target(tmp: &mut File, (buf_x, buf_y): (u32, u32), (target_x, target_y): (f64, f64)) {
+    let centre_x = (buf_x as f64 * target_x) as i64;
+    let centre_y = (buf_y as f64 * target_y) as i64;
+    let target_size = (max(buf_x, buf_y)/80) as i64;
+    use std::{io::Write};
+    let mut buf = std::io::BufWriter::new(tmp);
+    for y in 0..buf_y as i64 {
+        for x in 0..buf_x as i64 {
+            let distance_squared =(x-centre_x)*(x-centre_x) + (y-centre_y)*(y-centre_y);
+            if distance_squared >= target_size*target_size {
+                let a = 0xFF;
+                let r = 0x3F;
+                let g = 0x3F*y as u32/buf_y;
+                let b = 0x3F*y as u32/buf_y;
+                buf.write_all(&[b as u8, g as u8, r as u8, a as u8]).unwrap();
+            }
+            else {
+                let intensity = 0xFF - 0x40*(distance_squared/(target_size*target_size/4));
+                let a = 0xFF;
+                buf.write_all(&[intensity as u8, intensity as u8, intensity as u8, a as u8]).unwrap();
+            }
+        }
+    }
+    buf.flush().unwrap();
+}
+
+impl State {
+    fn draw_fullscreen_surfaces(&mut self, qh: &QueueHandle<State>) {
+
+        if self.compositor.is_none() || self.wm_base.is_none() || self.wl_shm.is_none() {
+            return;
+        }
+
+        let compositor = self.compositor.as_ref().unwrap();
+        let wm_base = self.wm_base.as_ref().unwrap();
+        let wl_shm = self.wl_shm.as_ref().unwrap();
+
+        for output in self.outputs.iter_mut() {
```

Comment:
```suggestion
        for (output, window) in self.outputs.iter_mut() {
```
would remove the awkward indexing into tuple happening later on.